### PR TITLE
Add type annotations

### LINKED ISSUE #183: Populate type signatures
Ensure that the signatures to all 'stactools.{core|cli|testing}' functions/methods have type annotations.

See #155 and #182

### PR DESCRIPTION
**Pull request by cuttlefish**
Closes #183 
`pystac 1.1.0` is required because it now publishes type information in its packages.

I had to add some error handling code to appease `mypy`, so I'd like to get this looked at carefully. I modified a lot of code I haven't dealt with before.

**PR checklist:**

- [x] Code is formatted (run `scripts/format`).
- [x] Code lints properly (run `scripts/lint`).
- [x] Tests pass (run `scripts/test`).
- [ ] Documentation has been updated to reflect changes, if applicable.
- [ ] Changes are added to the [CHANGELOG](../CHANGELOG.md).

### CODE REVIEW

File: src/stactools/core/io/__init__.py [hunk 1]
```diff
@@ -22,13 +22,13 @@ def read_text(href: str,
 class FsspecStacIO(DefaultStacIO):
     def read_text_from_href(self, href: str, *args: Any, **kwargs: Any) -> str:
         with fsspec.open(href, "r") as f:
-            return f.read()
+            return str(f.read())
```
Comment (gadomski) [inline src/stactools/core/io/__init__.py:25]:
```suggestion
            return str(f.read(), encoding='utf-8')
```

Ref: https://docs.python.org/3/library/stdtypes.html#str

File: src/stactools/core/io/xml.py [hunk 2]
```diff
@@ -52,8 +52,8 @@ def find_attr(self, attr: str, xpath: str) -> Optional[str]:
         return None if node is None else node.get_attr(attr)
 
     @property
-    def text(self):
-        return self.element.text
+    def text(self) -> Optional[str]:
+        return str(self.element.text)
```
Comment (gadomski) [inline src/stactools/core/io/xml.py:56]:
```suggestion
        if isinstance(self.element.text, bytes):
            return str(self.element.text, encoding='utf-8')
        elif isinstance(self.element.text, str):
            return str(self.element.text)
        else:
            assert (self.element.text == None)
            return None
```

**Review by gadomski — CHANGES_REQUESTED** [review]

File: src/stactools/core/io/__init__.py (hunk 1 above)
Comment (cuttlefish) [reply src/stactools/core/io/__init__.py]: Forcing the UTF8 decoding was causing tests to fail, so I did something similar to the XML handling change you suggested below.

**Comment by cuttlefish** [comment]
Thanks @gadomski! I wasn't sure how to handle the statically typed file inputs, but your suggestions look good.

**Review by gadomski — APPROVED** [review]

**Pull request merged**